unidist/core/backends/mpi/core/common.py

```diff
@@ -258,50 +258,76 @@ def get_op_name(op):
+        The actual object is for the data id owner, otherwise, ``None``.
     """
 
-    def __init__(self, id_value, garbage_collector):
-        super().__init__(id_value)
-        self._gc = garbage_collector if garbage_collector else None
+    _instances = weakref.WeakValueDictionary()
 
-    def __del__(self):
-        """Track object deletion by garbage collector."""
-        # We check for existence of `_qc` attribute because
-        # it might be deleted during serialization via `__getstate__`
-        if hasattr(self, "_gc") and self._gc is not None:
-            self._gc.collect(self.base_data_id())
+    def __new__(cls, owner_rank, data_number, gc=None):
+        key = (owner_rank, data_number)
+        if key in cls._instances:
+            return cls._instances[key]
+        else:
+            new_instance = super().__new__(cls)
+            cls._instances[key] = new_instance
+            return new_instance
 
-    def __getstate__(self):
-        """Remove a reference to garbage collector for correct `pickle` serialization."""
-        attributes = self.__dict__.copy()
-        # TODO: Look closer at the issue `KeyError: '_gc'` on small number of CPUs.
-        # There shouldn't be the error because workers should operate DataID(s) but not MasterDataID(s).
-        if hasattr(self, "_gc"):
-            del attributes["_gc"]
-        return attributes
-
-    def base_data_id(self):
+    def __init__(self, owner_rank, data_number, gc=None):
+        super().__init__(f"rank_{owner_rank}_id_{data_number}")
+        self.owner_rank = owner_rank
+        self.data_number = data_number
+        self._gc = gc
+
+    def __getnewargs__(self):
         """
-        Return the base class instance without garbage collector reference.
+        Prepare arguments to serialize and send this object to another MPI process.
 
         Returns
         -------
-        unidist.core.backends.common.data_id.DataID
-            Base ``DataID`` class object without garbage collector reference.
+        tuple
+            Tuple of the owner rank and data number.
         """
-        return DataID(self._id)
+        return (self.owner_rank, self.data_number)
+
+    def __getstate__(self):
+        """Remove a reference to garbage collector for correct `pickle` serialization."""
+        state = self.__dict__.copy()
+        # we remove this attribute for correct serialization,
+        # as well as to reduce the length of the serialized data
+        del state["_gc"]
+        return state
+
+    def __setstate__(self, state):
+        """
+        Add ``None`` reference for garbage collector for correct `pickle` deserialization.
+
+        Parameters
+        ----------
+        state : dict
+            The state of the object for deserialization.
+        """
+        state["_gc"] = None
+        self.__dict__ = state
+
+    def __del__(self):
+        """Track object deletion by garbage collector."""
+        if self._gc is not None:
```

Comment:
```suggestion
        if hasattr(self, "_gc") and self._gc is not None:
```
Also, let's put a comment why we check on existence of ._gc.